File: CPPScript/Получение ссылки на медиа-ресурс/Функция получения ссылки Youtube.cpp
// ------------------------------------------- Получение ссылки на Youtube ----
bool GetLink_Youtube31(string sLink) {
  string sData, sVideoID='', sMaxHeight='', sAudio='', sSubtitlesLanguage='ru';
  string sSubtitlesUrl, sFile, sVal, sMsg, sConfig, sHeaders, ttsDef; 
  TJsonObject JSON; TRegExpr RegEx;
  
  sHeaders = 'Referer: '+sLink+#13#10+
             'User-Agent: Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.100 Safari/537.36'+#13#10+
             'Origin: http://www.youtube.com'+#13#10;
  
  HmsRegExMatch('--maxheight=(\\d+)'    , mpPodcastParameters, sMaxHeight);
  HmsRegExMatch('--sublanguage=(\\w{2})', mpPodcastParameters, sSubtitlesLanguage);
  bool bSubtitles = (Pos('--subtitles'  , mpPodcastParameters)>0);  
  bool bAdaptive  = (Pos('--adaptive'   , mpPodcastParameters)>0);  
  bool bNotDE     = (Pos('notde=1'      , sLink)>0);  

  if (!HmsRegExMatch('[\\?&]v=([^&]+)'       , sLink, sVideoID))
  if (!HmsRegExMatch('youtu.be/([^&]+)'      , sLink, sVideoID))
       HmsRegExMatch('/(?:embed|v)/([^\\?]+)', sLink, sVideoID);

  if (sVideoID=='') return VideoMessage('Невозможно получить Video ID в ссылке Youtube');

  sLink = 'http://www.youtube.com/watch?v='+sVideoID+'&hl=ru&persist_hl=1&has_verified=1';
  
  sData = HmsDownloadURL(sLink, sHeaders, true);
  //sData = HmsUtf8Decode(sData);
  sData = HmsRemoveLineBreaks(sData);
  // Если еще не установлена реальная длительность видео - устанавливаем
  if ((Trim(mpTimeLength)=='') || (RightCopy(mpTimeLength, 6)=='00.000')) {
    if (HmsRegExMatch2('itemprop="duration"[^>]+content="..(\\d+)M(\\d+)S', sData, sVal, sMsg)) {
      PodcastItem[mpiTimeLength] = StrToInt(sVal)*60+StrToInt(sMsg);
    } else {
      sVal = HmsDownloadURL('http://www.youtube.com/get_video_info?html5=1&c=WEB&cver=html5&cplayer=UNIPLAYER&hl=ru&video_id='+sVideoID, sHeaders, true);
      if (HmsRegExMatch('length_seconds=(\\d+)', sData, sMsg))
        PodcastItem[mpiTimeLength] = StrToInt(sMsg);
    }
  }
  if (!HmsRegExMatch('player.config\\s*?=\\s*?({.*?});', sData, sConfig)) {
    // Если в загруженной странице нет нужной информации, пробуем немного по-другому
    sLink = 'http://hms.lostcut.net/youtube/g.php?v='+sVideoID;
    if (sMaxHeight!=''                  ) sLink += '&max_height='+sMaxHeight;
    if (Trim(mpPodcastMediaFormats )!='') sLink += '&media_formats='+mpPodcastMediaFormats;
    if (bAdaptive                       ) sLink += '&adaptive=1';
    sData = HmsUtf8Decode(HmsDownloadUrl(sLink));
    if (HmsRegExMatch('"reason":"(.*?)"' , sData, sMsg)) { 
      HmsLogMessage(2 , sMsg); 
      VideoMessage(sMsg); 
      return; 
    } else {
      sData = HmsJsonDecode(sData);
      HmsRegExMatch('"url":"(.*?)"', sData, MediaResourceLink);
      return true;
    }
  }
  
  String hlsUrl, ttsUrl, flp, jsUrl, dashMpdLink, streamMap, playerId, algorithm;
  String sType, itag, sig, alg, s;
  String UrlBase = "";
  int  i, n, w, num, height, priority, minPriority = 90, selHeight, maxHeight = 1080;
  bool is3D; 
  TryStrToInt(sMaxHeight, maxHeight);
  JSON = TJsonObject.Create();
  try {
    JSON.LoadFromString(sConfig);
    hlsUrl      = HmsExpandLink(JSON.S['args\\hlsvp' ], UrlBase);
    ttsUrl      = HmsExpandLink(JSON.S['args\\caption_tracks'], UrlBase);
    flp         = HmsExpandLink(JSON.S['url'         ], UrlBase);
    jsUrl       = HmsExpandLink(JSON.S['assets\\js'  ], UrlBase);
    streamMap   = JSON.S['args\\url_encoded_fmt_stream_map'];
    if (bAdaptive && JSON.B['args\\adaptive_fmts']) 
      streamMap = JSON.S['args\\adaptive_fmts'];
    if ((streamMap=='') && (hlsUrl=='')) {
      sMsg = "Невозможно найти данные для воспроизведения на странице видео.";
      if (HmsRegExMatch('(<h\\d[^>]+class="message".*?</h\\d>)', sData, sMsg)) sMsg = HmsUtf8Decode(HmsHtmlToText(sMsg));
      HmsLogMessage(2, sMsg);
      VideoMessage(sMsg); 
      return;
    }
  } finally { JSON.Free; }
  if (Copy(jsUrl, 1, 2)=='//') jsUrl = 'http:'+Trim(jsUrl);
  HmsRegExMatch('/player-([\\w_-]+)/', jsUrl, playerId);
  algorithm = HmsDownloadURL('https://hms.lostcut.net/youtube/getalgo.php?jsurl='+HmsHttpEncode(jsUrl));
  
  if (hlsUrl!='') {
    MediaResourceLink = ' '+hlsUrl;

    sData = HmsDownloadUrl(sLink, sHeaders, true);
    RegEx = TRegExpr.Create('BANDWIDTH=(\\d+).*?RESOLUTION=(\\d+)x(\\d+).*?(http[^#]*)', PCRE_SINGLELINE);
    try {
      if (RegEx.Search(sData)) do {
        sLink = '' + RegEx.Match(4);
        height = StrToIntDef(RegEx.Match(3), 0);
        if (mpPodcastMediaFormats!='') {
          priority = HmsMediaFormatPriority(height, mpPodcastMediaFormats);
          if ((priority>=0) && (priority>minPriority)) {
            MediaResourceLink = sLink; minPriority = priority;
          }
        } else if ((height > selHeight) && (height <= maxHeight)) {
          MediaResourceLink = sLink; selHeight = height;
        }
      } while (RegEx.SearchAgain());
    } finally { RegEx.Free(); }
  
  } else if (streamMap!='') {
      i=1; while (i<=Length(streamMap)) {
        sData = Trim(ExtractStr(streamMap, ',', i));
        sType = HmsHttpDecode(ExtractParam(sData, 'type', '', '&'));
        itag  = ExtractParam(sData, 'itag'    , '', '&');
        is3D  = ExtractParam(sData, 'stereo3d', '', '&') == '1';
        sLink = '';
        if (Pos('url=', sData)>0) {
          sLink = ' ' + HmsHttpDecode(ExtractParam(sData, 'url', '', '&'));
          if (Pos('&signature=', sLink)==0) {
            sig = HmsHttpDecode(ExtractParam(sData, 'sig', '', '&'));    
            if (sig=='') {
              sig = HmsHttpDecode(ExtractParam(sData, 's', '', '&'));
              for (w=1; w<=WordCount(algorithm, ' '); w++) {
                alg = ExtractWord(w, algorithm, ' ');
                if (Length(alg)<1) continue;
                if (Length(alg)>1) TryStrToInt(Copy(alg, 2, 4), num);
                if (alg[1]=='r') {s=''; for(n=Length(sig); n>0; n--) s+=sig[n]; sig = s;   } // Reverse
                if (alg[1]=='s') {sig = Copy(sig, num+1, Length(sig));                     } // Clone
                if (alg[1]=='w') {n = (num-Trunc(num/Length(sig)))+1; Swap(sig[1], sig[n]);} // Swap
              }
            }
            if (sig!='') sLink += '&signature=' + sig;
          }
        }
        if (itag in ([139,140,141,171,172])) { sAudio = sLink; continue; }
        if (sLink!='') {
          height = 0; //http://www.genyoutube.net/formats-resolution-youtube-videos.html
            if      (itag in ([13,17,160                  ])) height = 144;
            else if (itag in ([5,36,92,132,133,242        ])) height = 240;
            else if (itag in ([6                          ])) height = 270;
            else if (itag in ([18,34,43,82,100,93,134,243 ])) height = 360;
            else if (itag in ([35,44,83,101,94,135,244,43 ])) height = 480;
            else if (itag in ([22,45,84,102,95,136,298,247])) height = 720;
            else if (itag in ([37,46,85,96,137,248,299    ])) height = 1080;
            else if (itag in ([264,271                    ])) height = 1440;
            else if (itag in ([266,138                    ])) height = 2160;
            else if (itag in ([272                        ])) height = 2304;
            else if (itag in ([38                         ])) height = 3072;
            else continue;
            if (mpPodcastMediaFormats!='') {
              priority = HmsMediaFormatPriority(height, mpPodcastMediaFormats);
              if ((priority>=0) || (priority<minPriority)) {
                MediaResourceLink = sLink; minPriority = priority; selHeight = height;
              }
            } else if ((height>selHeight) && (height<= maxHeight)) {
              MediaResourceLink = sLink; selHeight = height;
            
            } else if ((height>=selHeight) && (height<= maxHeight) && (itag in ([18,22,37,38,82,83,84,85]))) {
              // Если выоста такая же, но формат MP4 - то выбираем именно его (делаем приоритет MP4)
              MediaResourceLink = sLink; selHeight = height;
            }
        }
      }
      if (bAdaptive && (sAudio!='')) MediaResourceLink = '-i "'+Trim(MediaResourceLink)+'" -i "'+Trim(sAudio)+'"';
    
  }
  // Если есть субтитры и в дополнительных параметрах указано их показывать - загружаем 
  if (bSubtitles && (ttsUrl!='')) {
    string sTime1, sTime2; float nStart, nDur;
    sLink = ''; n = WordCount(ttsUrl, ',');
    for (i=1; i <= n; i++) {
      sData = ExtractWord(i, ttsUrl, ',');
      sType = HmsPercentDecode(ExtractParam(sData, 'lc', '', '&'));
      sVal  = HmsPercentDecode(ExtractParam(sData, 'u' , '', '&'));
      if (sType == 'en') sLink = sVal;
      if (sType == sSubtitlesLanguage) { sLink = sVal; break; }
    }
    if (sLink != '') {
      sData = HmsDownloadURL(sLink, sHeaders, true);
      sMsg  = ''; i = 0;
      RegEx = TRegExpr.Create('(<(text|p).*?</(text|p)>)', PCRE_SINGLELINE); // Convert to srt format
      try {
        if (RegEx.Search(sData)) do {
          if      (HmsRegExMatch('start="([\\d\\.]+)', RegEx.Match, sVal)) nStart = StrToFloat(ReplaceStr(sVal, '.', ','))*1000;
          else if (HmsRegExMatch('t="(\\d+)'         , RegEx.Match, sVal)) nStart = StrToFloat(sVal);
          if      (HmsRegExMatch('dur="([\\d\\.]+)'  , RegEx.Match, sVal)) nDur   = StrToFloat(ReplaceStr(sVal, '.', ','))*1000;
          else if (HmsRegExMatch('d="(\\d+)'         , RegEx.Match, sVal)) nDur   = StrToFloat(sVal);
          sTime1 = HmsTimeFormat(Int(nStart/1000))+','+RightCopy(Str(nStart), 3);
          sTime2 = HmsTimeFormat(Int((nStart+nDur)/1000))+','+RightCopy(Str(nStart+nDur), 3);
          sMsg += Format("%d\n%s --> %s\n%s\n\n", [i, sTime1, sTime2, HmsHtmlToText(HmsHtmlToText(RegEx.Match(0), 65001))]);
          i++;
        } while (RegEx.SearchAgain());
      } finally { RegEx.Free(); }
      sFile = HmsSubtitlesDirectory+'\\Youtube\\'+PodcastItem.ItemID+'.'+sSubtitlesLanguage+'.srt';
      HmsStringToFile(sMsg, sFile);
      PodcastItem[mpiSubtitleLanguage] = sFile;
    }
  }
}

File: PascalScript/Получение ссылки на медиа-ресурс/Функция получения ссылки Youtube.pas
// ---- Получение ссылки на Youtube -------------------------------------------
Function GetLink_Youtube31(sLink: String): Boolean;
Var
  sData, sVideoID, sMaxHeight, sSubtitlesLanguage:String;
  hlsUrl, ttsUrl, flp, jsUrl, dashMpdLink, streamMap, playerId, algorithm: String;
  sType, itag, sig, alg, s, sFile, sVal, sConfig, sHeaders: String ;
  UrlBase:String = ""; is3D, bSubtitles, bAdaptive: Boolean;
  i, n, w, num, height, priority, minPriority, selHeight, maxHeight: Integer;
  JSON: TJsonObject; RegEx: TRegExpr;
Begin 
  sSubtitlesLanguage :='ru'; minPriority := 90; maxHeight := 1080;
  sHeaders := 'Referer: '+sLink+#13#10+
              'User-Agent: Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/46.0.2490.86 Safari/537.36'+#13#10+
              'Origin: http://www.youtube.com'+#13#10;
  
  HmsRegExMatch('--maxheight=(\d+)'    , mpPodcastParameters, sMaxHeight);
  HmsRegExMatch('--sublanguage=(\w{2})', mpPodcastParameters, sSubtitlesLanguage);
  bSubtitles := (Pos('--subtitles'  , mpPodcastParameters)>0);  
  bAdaptive  := (Pos('--adaptive'   , mpPodcastParameters)>0);  
  TryStrToInt(sMaxHeight, maxHeight);

  If Not HmsRegExMatch('[\?&]v=([^&]+)'       , sLink, sVideoID) Then
         HmsRegExMatch('/(?:embed|v)/([^\?]+)', sLink, sVideoID);
  
  If sVideoID='' Then Begin
    sData := HmsDownloadURL(sLink, sHeaders, True);
    If Not HmsRegExMatch('youtube.com[^"''>]+v=([^&]+)', sData, sVideoID) Then
           HmsRegExMatch('/embed/([^\?]+)'             , sData, sVideoID);
  End;

  If sVideoID='' Then Begin HmsLogMessage(2, 'Невозможно получить Video ID в ссылке Youtube'); Exit; End;

  sLink := 'http://www.youtube.com/watch?v='+sVideoID+'&hl=ru&persist_hl=1&has_verified=1';
  
  sData := HmsUtf8Decode(HmsDownloadURL(sLink));
  sData := HmsRemoveLineBreaks(sData);
  HmsRegExMatch('player.config\s*?=\s*?({.*?});', sData, sConfig);
  
  JSON := TJsonObject.Create();
  Try
    JSON.LoadFromString(sConfig);
    hlsUrl      := HmsExpandLink(JSON.S['args\hlsvp' ], UrlBase);
    ttsUrl      := HmsExpandLink(JSON.S['args\ttsurl'], UrlBase);
    flp         := HmsExpandLink(JSON.S['url'        ], UrlBase);
    jsUrl       := HmsExpandLink(JSON.S['assets\js'  ], UrlBase);
    streamMap   := JSON.S['args\url_encoded_fmt_stream_map'];
    If bAdaptive AND JSON.B['args\adaptive_fmts'] Then 
      streamMap := JSON.S['args\adaptive_fmts'];
    If (streamMap='') AND (hlsUrl='') Then Begin
      HmsLogMessage(2, "Can not found stream map in player config");
      Exit;
    End;
  Finally 
    JSON.Free;
  End;
  If Copy(jsUrl, 1, 2)='//' Then jsUrl := 'http:'+Trim(jsUrl);
  HmsRegExMatch('/player-([\w_-]+)/', jsUrl, playerId);
  algorithm := HmsDownloadURL('https://hms.lostcut.net/youtube/getalgo.php?jsurl='+HmsHttpEncode(jsUrl));
  
  If hlsUrl<>'' Then Begin
    MediaResourceLink := ' '+hlsUrl;
    sData := HmsDownloadUrl(sLink, sHeaders, true);
    RegEx := TRegExpr.Create('BANDWIDTH=(\d+).*?RESOLUTION=(\d+)x(\d+).*?(http[^#]*)', PCRE_SINGLELINE);
    try 
      If RegEx.Search(sData) Then Repeat
        sLink  := ' ' + RegEx.Match(4);
        height := StrToIntDef(RegEx.Match(3), 0);
        If mpPodcastMediaFormats<>'' Then Begin
          priority := HmsMediaFormatPriority(height, mpPodcastMediaFormats);
          If (priority>=0) AND (priority>minPriority) Then Begin
            MediaResourceLink := sLink; minPriority := priority;
          End;
        End Else If ((height > selHeight) AND (height <= maxHeight)) Then Begin
          MediaResourceLink := sLink; selHeight := height;
        End;
      Until Not RegEx.SearchAgain();
    Finally 
      RegEx.Free();
    End;
  
  End Else If streamMap<>'' Then Begin
    i:=1; While (i<=Length(streamMap)) Do 
    Begin
      sData := Trim(ExtractStr(streamMap, ',', i));
      sType := HmsHttpDecode(ExtractParam(sData, 'type', '', '&'));
      itag  := ExtractParam(sData, 'itag'    , '', '&');
      is3D  := ExtractParam(sData, 'stereo3d', '', '&') = '1';
      sLink := '';
      If (Pos('url=', sData)>0) Then Begin
        sLink := ' ' + HmsHttpDecode(ExtractParam(sData, 'url', '', '&'));
        If (Pos('&signature=', sLink)=0) Then Begin
          sig := HmsHttpDecode(ExtractParam(sData, 'sig', '', '&'));    
          If (sig='') Then Begin
            sig := HmsHttpDecode(ExtractParam(sData, 's', '', '&'));
            For w:=1 To WordCount(algorithm, ' ') Do Begin
              alg := ExtractWord(w, algorithm, ' ');
              If (Length(alg)<1) Then continue;
              If (Length(alg)>1) Then TryStrToInt(Copy(alg, 2, 4), num);
              If (alg[1]='r') Then Begin s:=''; for n:=Length(sig) DownTo 1 Do s:=s+sig[n]; sig := s; End; // Reverse
              If (alg[1]='s') Then Begin sig := Copy(sig, num+1, Length(sig));                        End; // Clone
              If (alg[1]='w') Then Begin n := (num-Trunc(num/Length(sig)))+1; Swap(sig[1], sig[n]);   End; // Swap
            End;
          End;
          If sig<>'' Then sLink := sLink + '&signature=' + sig;
        End;
      End;
      If (sLink<>'') Then Begin
        If ((Pos('flv', sType)>0) Or (Pos('mp4', sType)>0)) Then Begin
          height := 0;
          If      (itag in ([13,17,160,36           ])) Then height := 144
          Else if (itag in ([5,83,133,242           ])) Then height := 240
          Else if (itag in ([6                      ])) Then height := 270
          Else if (itag in ([18,34,43,82,100,134,243])) Then height := 360
          Else if (itag in ([35,44,101,135,244,43   ])) Then height := 480
          Else if (itag in ([22,45,84,102,136,247   ])) Then height := 720
          Else if (itag in ([37,46,137,248          ])) Then height := 1080
          Else if (itag in ([264,271                ])) Then height := 1440
          Else if (itag in ([266                    ])) Then height := 2160
          Else if (itag in ([138,272                ])) Then height := 2304
          Else if (itag in ([38                     ])) Then height := 3072;
          If (mpPodcastMediaFormats<>'') Then Begin
            priority := HmsMediaFormatPriority(height, mpPodcastMediaFormats);
            If ((priority>=0) Or (priority<minPriority)) Then Begin
              MediaResourceLink := sLink; minPriority := priority; selHeight := height;
            End;
          End Else If ((height>selHeight) And (height<= maxHeight)) Then Begin
            MediaResourceLink := sLink; selHeight := height;
          End;
        End;
      End;
    End;
    
  End;
  // Если есть субтитры и в дополнительных параметрах указано их показывать - загружаем 
  If (bSubtitles AND (ttsUrl<>'')) Then Begin
    sFile := HmsSubtitlesDirectory+'\Youtube\'+PodcastItem.ItemID+'.'+sSubtitlesLanguage+'.srt';
    sLink := ttsUrl+'&fmt=srt&lang='; 
    If Not HmsDownloadURLToFile(sLink+sSubtitlesLanguage, sFile, 'Accept-Encoding: gzip, deflate') Then
           HmsDownloadURLToFile(sLink+'en'              , sFile, 'Accept-Encoding: gzip, deflate');
    PodcastItem[mpiSubtitleLanguage] := sFile;
  End;
End;